File: 2015/Day06_2015.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Day06_2015
{
    public static void main(String[] args) throws FileNotFoundException
    {
        // Guaranteed to initialize to 0
        int[][] grid = new int[1000][1000];
        int[][] grid2 = new int[1000][1000];

        File file = new File("Day06_2015.txt");
        Scanner scnr = new Scanner(file);

        while (scnr.hasNextLine())
        {
            String[] xList = scnr.nextLine().split("\\s");
            
            String[] xy = xList[xList.length - 3].split(",");
            int x1 = Integer.parseInt(xy[0]);
            int y1 = Integer.parseInt(xy[1]);
            String[] xy2 = xList[xList.length - 1].split(",");
            int x2 = Integer.parseInt(xy2[0]) + 1;
            int y2 = Integer.parseInt(xy2[1]) + 1;
            
            if (xList.length == 5)
            {
                if (xList[1].equals("on"))
                {
                    for (int i = x1; i < x2; i++)
                    {
                        for (int j = y1; j < y2; j++)
                        {
                            grid[i][j] = 1;
                            grid2[i][j]++;
                        }
                    }
                }
                else if (xList[1].equals("off"))
                {
                    for (int i = x1; i < x2; i++)
                    {
                        for (int j = y1; j < y2; j++)
                        {
                            grid[i][j] = 0;
                            if (grid2[i][j] > 0)
                                grid2[i][j]--;
                        }
                    }
                }
                else
                {
                    System.out.println("Error: 5list[1] returned incorrect string");
                    scnr.close();
                    return;
                }
            }
            else if (xList.length == 4)
            {
                for (int i = x1; i < x2; i++)
                {
                    for (int j = y1; j < y2; j++)
                    {
                        if (grid[i][j] == 0)
                            grid[i][j] = 1;
                        else if (grid[i][j] == 1)
                            grid[i][j] = 0;
                        else
                        {
                            System.out.println("Error: grid value not 0 or 1");
                            scnr.close();
                            return;
                        }

                        grid2[i][j] += 2;
                    }
                }
            }
            else
            {
                System.out.println("Error: bad list size returned");
                scnr.close();
                return;
            }
        }

        scnr.close();

        int lightCount = 0;
        int brightness = 0;

        for (int i = 0; i < 1000; i++)
        {
            for (int j = 0; j < 1000; j++)
            {
                if (grid[i][j] == 1)
                    lightCount++;
                brightness += grid2[i][j];
            }
        }
        
        System.out.println("Total number of lights turned on: " + lightCount);
        System.out.println("Cumulative light brightness: " + brightness);
    }
}

File: 2015/Day06_2015.cpp
#include <iostream>
#include <fstream>
#include <string>

std::string* split(std::string& s, int size, char delim);

int main()
{
    const int xMax = 1000;
    const int yMax = 1000;

    std::ifstream file("Day06_2015.txt");
    std::string line;

    int* grid = new int[xMax * yMax]{0};
    int* grid2 = new int[xMax * yMax]{0};

    while (std::getline(file, line))
    {
        int numElements = 0;

        if (line[1] == 'u')
            numElements = 5;
        else if (line[1] == 'o')
            numElements = 4;

        std::string* xList = split(line, numElements, ' ');

        std::string* xy = split(xList[numElements - 3], 2, ',');
        int x1 = std::stoi(xy[0]);
        int y1 = std::stoi(xy[1]);
        std::string* xy2 = split(xList[numElements - 1], 2, ',');
        int x2 = std::stoi(xy2[0]) + 1;
        int y2 = std::stoi(xy2[1]) + 1;

        if (xy != nullptr)
            delete[] xy;
        if (xy2 != nullptr)
            delete[] xy2;

        if (numElements == 5)
        {
            if (xList[1] == "on")
            {
                for (int i = x1; i < x2; i++)
                {
                    for (int j = y1; j < y2; j++)
                    {
                        grid[xMax * i + j] = 1;
                        grid2[xMax * i + j]++;
                    }
                }
            }
            else if (xList[1] == "off")
            {
                for (int i = x1; i < x2; i++)
                {
                    for (int j = y1; j < y2; j++)
                    {
                        grid[xMax * i + j] = 0;
                        if (grid2[xMax * i + j] > 0)
                            grid2[xMax * i + j]--;
                    }
                }
            }
            else
            {
                std::cout << "Error: 5list[1] returned incorrect string\n";
                file.close();
                if (xList != nullptr)
                    delete[] xList;
                if (grid != nullptr)
                    delete[] grid;
                if (grid2 != nullptr)
                    delete[] grid2;
                return -1;
            }
        }
        else if (numElements == 4)
        {
            for (int i = x1; i < x2; i++)
            {
                for (int j = y1; j < y2; j++)
                {
                    if (grid[xMax * i + j] == 0)
                        grid[xMax * i + j] = 1;
                    else if (grid[xMax * i + j] == 1)
                        grid[xMax * i + j] = 0;
                    else
                    {
                        std::cout << "Error: grid value not 0 or 1\n";
                        file.close();
                        if (xList != nullptr)
                            delete[] xList;
                        if (grid != nullptr)
                            delete[] grid;
                        if (grid2 != nullptr)
                            delete[] grid2;
                        return -1;
                    }

                    grid2[xMax * i + j] += 2;
                }
            }
        }
        else
        {
            std::cout << "Error: bad list size returned\n";
            file.close();
            if (xList != nullptr)
                delete[] xList;
            if (grid != nullptr)
                delete[] grid;
            if (grid2 != nullptr)
                delete[] grid2;
            return -1;
        }

        if (xList != nullptr)
            delete[] xList;
    }

    file.close();

    int lightCount = 0;
    int brightness = 0;

    for (int i = 0; i < (xMax * yMax); i++)
    {
        if (grid[i] == 1)
            lightCount++;
        brightness += grid2[i];
    }
        
    std::cout << "Total number of lights turned on: " << lightCount << "\n";
    std::cout << "Cumulative light brightness: " << brightness << "\n";

    if (grid != nullptr)
        delete[] grid;
    if (grid2 != nullptr)
        delete[] grid2;

    return 0;
}

// splits a string into a string array based on a given char
// delimiter. Could be generalized with minimal alteration.
// NOTE: memory handled by CALLER.
std::string* split(std::string& s, int size, char delim)
{
    std::string* lineElements = new std::string[size];
    
    size_t last = 0;
    size_t next = s.find(delim);
    int index = 0;

    while (next != std::string::npos)
    {
        if (index >= size)
            throw("Index out of range.");

        lineElements[index] = s.substr(last, next - last);
        last = next + 1;
        next = s.find(delim, last);

        index++;
    }

    if (index < size)
        lineElements[index] = s.substr(last, next); // next should be npos which indicates all characters to end of string
    else
        throw("Index out of range.");
    
    return lineElements;
}